### PR TITLE
Types

### LINKED ISSUE #453: Typescript Typings
Hi guys, awesome work, can you please add typescript support ?

I'm working on a project using typescript, but you have not support for it :(

### PR DESCRIPTION
**Pull request by rgomezp**
Types for the React Native Major-Release

Closes #453

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,395 @@
+declare module 'react-native-onesignal' {
+    /* O P T I O N   T Y P E   V A L U E S */
```
Comment (jkasten2) [inline src/index.d.ts:2]: The spaces between each letter means this won't be found in a search.

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    export type OpenedEventActionType = 0 | 1;
+
+    /* O B S E R V E R  C H A N G E  E V E N T S */
+    export interface ChangeEvent<T> {
+        from : T;
+        to   : T;
+    }
+
+    export interface PermissionChange {
+        status                  ?: IosPermissionStatus;    // ios
```
Comment (jkasten2) [inline src/index.d.ts:19]: The alignment here is interesting, is there a specific standard you are following here?

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        hasPrompted             ?: boolean;   // ios
+        provisional             ?: boolean;   // ios
+        areNotificationsEnabled ?: boolean;   // android
+    };
+
+    export interface SubscriptionChange {
+        userId                  : string;
```
Comment (jkasten2) [inline src/index.d.ts]: Can be `null`

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+        pushToken               : string;
```
Comment (jkasten2) [inline src/index.d.ts]: Can be `null`

File: src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+    export interface EmailSubscriptionChange {
+        emailAddress        ?: string;
+        emailUserId         ?: string;
+        isEmailSubscribed   : boolean;
+    };
+
+    /* N O T I F I C A T I O N S */
+    export interface OSNotification {
+        body            : string;
+        sound           : string;
```
Comment (jkasten2) [inline src/index.d.ts]: Can be `null`

File: src/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+        title           : string;
```
Comment (jkasten2) [inline src/index.d.ts]: Can be `null`

File: src/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        launchURL       : string;
```
Comment (jkasten2) [inline src/index.d.ts]: Can be `null`

File: src/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+        rawPayload      : object | string;
```
Comment (jkasten2) [inline src/index.d.ts]: I would expect this to just be one type always.

File: src/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+        actionButtons   : object[];
```
Comment (jkasten2) [inline src/index.d.ts]: Can be `null`

File: src/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        // ios only
+        badge               ?: string;
+        category            ?: string;
+        threadId            ?: string;
+        subtitle            ?: string;
+        templateId          ?: string;
+        attachments         ?: object;
+        templateName        ?: string;
+        mutableContent      ?: boolean;
+        badgeIncrement      ?: string;
```
Comment (jkasten2) [inline src/index.d.ts]: nit, group `badgeIncrement` to be next to `badge`

File: src/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+        lockScreenVisibility    ?: string;
+        androidNotificationId   ?: number;
+        // ios only
+        badge               ?: string;
+        category            ?: string;
+        threadId            ?: string;
+        subtitle            ?: string;
+        templateId          ?: string;
+        attachments         ?: object;
+        templateName        ?: string;
```
Comment (jkasten2) [inline src/index.d.ts:68]: nit, group template fields together

File: src/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+        notificationId  : string;
+        // android only
+        groupKey                ?: string;
+        ledColor                ?: string;
+        priority                ?: number;
+        smallIcon               ?: string;
+        largeIcon               ?: string;
+        bigPicture              ?: string;
+        collapseId              ?: string;
+        groupMessage            ?: string;
```
Comment (jkasten2) [inline src/index.d.ts]: nit, group `group` and `groupMessage` together

File: src/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+         * @returns void
+         */
+        setNotificationOpenedHandler(handler: (openedEvent: OpenedEvent) => void): void;
+
+        /**
+         * Prompts the iOS user for push notifications.
+         * @param  {(response:boolean)=>void} handler
+         * @returns void
+         */
+        promptForPushNotificationsWithUserResponse(handler: (response: boolean) => void): void;
```
Comment (jkasten2) [inline src/index.d.ts]: `handler` should be optional

File: src/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+        deleteTags(keys: string[]);
+
+        /**
+         * Allows you to set the user's email address with the OneSignal SDK.
+         * @param  {string} email
+         * @param  {string} authCode
+         * @param  {Function} handler
+         * @returns void
+         */
+        setEmail(email: string, authCode: string, handler: Function): void;
```
Comment (jkasten2) [inline src/index.d.ts]: `authCode` and `handler` should be optional

File: src/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+
+        /**
+         * If your app implements logout functionality, you can call logoutEmail to dissociate the email from the device.
+         * @param  {Function} handler
+         */
+        logoutEmail(handler: Function);
```
Comment (jkasten2) [inline src/index.d.ts]: `handler` is optional

File: src/index.d.ts [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+         * Send a notification
+         * @param  {string} notificationObjectString - JSON string payload (see REST API reference)
+         * @param  {(success:object)=>void} onSuccess
+         * @param  {(failure:object)=>void} onFailure
+         * @returns void
+         */
+        postNotification(notificationObjectString: string, onSuccess: (success: object) => void, onFailure: (failure: object) => void): void;
+
+        /**
+         * iOS provides a standard way to clear notifications by clearing badge count.
```
Comment (jkasten2) [inline src/index.d.ts]: I believe we changed this wording in the OneSignal public docs. It should match that

File: src/index.d.ts [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+         * @returns void
+         */
+        clearOneSignalNotifications(): void;
+
+        /**
+         * Cancels a single OneSignal notification based on its Android notification integer id.
+         * @returns void
+         */
+        cancelNotification(): void;
```
Comment (jkasten2) [inline src/index.d.ts]: Missing required `androidNotificationId: number`

File: src/index.d.ts [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+
+        /**
+         * Allows you to use your own system's user ID's to send push notifications to your users.
+         * @param  {string} externalId
+         * @param  {(results:object)=>void} handler
+         * @returns void
+         */
+        setExternalUserId(externalId: string, handler: (results: object) => void): void;
```
Comment (jkasten2) [inline src/index.d.ts]: `handler` not required

File: src/index.d.ts [hunk 19]
```diff
@@ -0,0 +1,395 @@
...
+
+        /**
+         * Removes whatever was set as the current user's external user ID.
+         * @param  {(results:object)=>void} handler
+         * @returns void
+         */
+        removeExternalUserId(handler: (results: object) => void): void;
```
Comment (jkasten2) [inline src/index.d.ts]: `handler` not required

**Review by jkasten2 — CHANGES_REQUESTED** [review]

File: src/index.d.ts (hunk 1 above)
Comment (rgomezp) [reply src/index.d.ts:2]: these are meant as organizational comments only. makes it easy to see what the different sections are

File: src/index.d.ts (hunk 2 above)
Comment (rgomezp) [reply src/index.d.ts:19]: I've used this standard elsewhere. It is meant to make the code more readable and quickly scannable

File: src/index.d.ts (hunk 16 above)
Comment (rgomezp) [reply src/index.d.ts]: [From docs](https://documentation.onesignal.com/docs/ios-native-sdk#clearonesignalnotifications):

> iOS provides a standard way to clear notifications by clearing badge count, thus there is no specific OneSignal API call for clearing notifications.

File: src/index.d.ts (hunk 17 above)
Comment (rgomezp) [reply src/index.d.ts]: good catch

File: src/index.d.ts (hunk 19 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 18 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 17 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 15 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 5 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 6 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 7 above)
Comment (rgomezp) [reply src/index.d.ts:43]: Done.

File: src/index.d.ts (hunk 9 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 10 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 11 above)
Comment (rgomezp) [reply src/index.d.ts:68]: Done.

File: src/index.d.ts (hunk 12 above)
Comment (rgomezp) [reply src/index.d.ts]: groupKey*

Done.

File: src/index.d.ts (hunk 13 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 14 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 4 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 3 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 8 above)
Comment (rgomezp) [reply src/index.d.ts]: this would require a change in the bridges as they return different types across iOS & Android

File: src/index.d.ts (hunk 2 above)
Comment (rgomezp) [reply src/index.d.ts:19]: [More info on this standard](https://www.freecodecamp.org/news/how-to-columnize-your-code-to-improve-readability-f1364e2e77ba/)

File: src/index.d.ts (hunk 16 above)
Comment (jkasten2) [reply src/index.d.ts]: I was thinking this one as it does indicate it is only for Android.
> Android only. iOS provides a standard way to clear notifications by clearing badge count. There is no specific OneSignal API call for clearing notifications.
https://documentation.onesignal.com/docs/sdk-reference#clearonesignalnotifications-method

I am ok with either description for now in this PR but thinking we just need to add "only Android" for sure.

File: src/index.d.ts (hunk 8 above)
Comment (rgomezp) [reply src/index.d.ts]: we can change this later on in its own PR

File: src/index.d.ts [hunk 20]
```diff
@@ -267,24 +267,25 @@ declare module 'react-native-onesignal' {
 
         /**
          * Cancels a single OneSignal notification based on its Android notification integer id.
+         * @param  {string} id - notification id to cancel
```
Comment (rgomezp) [inline src/index.d.ts]: Add "Android only"

Should be `number`

Comment (rgomezp) [reply src/index.d.ts]: Done.

File: src/index.d.ts (hunk 16 above)
Comment (rgomezp) [reply src/index.d.ts]: Done.

**Pull request merged**